analyzers/tests/SonarAnalyzer.Test/Wrappers/RegisterSymbolStartActionWrapperTest.cs

```diff
@@ -0,0 +1,319 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.ShimLayer.AnalysisContext;
+using CS = Microsoft.CodeAnalysis.CSharp;
+using SonarSymbolStartAnalysisContext = SonarAnalyzer.ShimLayer.AnalysisContext.SymbolStartAnalysisContext;
+using VB = Microsoft.CodeAnalysis.VisualBasic;
+
+namespace SonarAnalyzer.Test.Wrappers;
+
+[TestClass]
+public class RegisterSymbolStartActionWrapperTest
+{
+    [TestMethod]
+    public async Task RegisterSymbolStartAction_SymbolStartProperties()
+    {
+        var code = """
+            public class C
+            {
+                int i = 0;
+                public void M() => ToString();
+            }
+            """;
+        var snippet = new SnippetCompiler(code);
+        var symbolStartWasCalled = false;
+        var compilation = snippet.Compilation.WithAnalyzers(ImmutableArray.Create<DiagnosticAnalyzer>(
+            new TestDiagnosticAnalyzer(symbolStart =>
+            {
+                symbolStart.CancellationToken.IsCancellationRequested.Should().BeFalse();
+                symbolStart.Compilation.SyntaxTrees.Should().ContainSingle();
+                symbolStart.Options.Should().NotBeNull();
+                symbolStart.Symbol.Should().BeAssignableTo<INamedTypeSymbol>().Which.Name.Should().Be("C");
+                symbolStartWasCalled = true;
+            }, SymbolKind.NamedType)));
+        var diags = await compilation.GetAnalyzerDiagnosticsAsync();
+        symbolStartWasCalled.Should().BeTrue();
+        diags.Should().BeEmpty();
+    }
+
+    [TestMethod]
+    public async Task RegisterSymbolStartAction_RegisterCodeBlockAction()
+    {
+        var code = """
```

Comment:
I was thinking that it would be good to extend the code snippet to test (and showcase) how per-symbol registration can be used to limit the scope. Something like the following:
```cs
var code = """
    public class C
    {
        int i = 0;
        public void M() => ToString();
    }
    public class D
    {
        int j = 0;
        public void N() => ToString();
    }
    """;
var snippet = new SnippetCompiler(code);
var visitedCodeBlocks = new List<string>();
var compilation = snippet.Compilation.WithAnalyzers(ImmutableArray.Create<DiagnosticAnalyzer>(
    new TestDiagnosticAnalyzer(symbolStart =>
    {
        if (symbolStart.Symbol.Name == "C") return;

        symbolStart.RegisterCodeBlockAction(block =>
        {
            var node = block.CodeBlock.ToString();
            visitedCodeBlocks.Add(node);
        });
    }, SymbolKind.NamedType)));
await compilation.GetAnalyzerDiagnosticsAsync();
visitedCodeBlocks.Should().BeEquivalentTo("int j = 0;", "public void N() => ToString();");
```

What do you think?